### PR TITLE
Retry liveness on zookeeper check

### PR DESCRIPTION
**Pull request by tor-vs-floki**
Internal ticket 175

### CODE REVIEW

File: bubuku/features/restart_if_dead.py [hunk 1]
```diff
@@ -18,18 +19,31 @@ def __init__(self, broker: BrokerManager, zk: BukuExhibitor):
...
 
         return RestartBrokerChange(self.zk, self.broker, _cancel_if, self.on_check_removed)
 
+    def is_running_and_registered(self, tries=1):
+        if not self.broker.process.is_running():
+            return False
+        # Retry with a sleep of 2 seconds in between
+        for x in range(0, tries):
+            if (x > 0):
+                sleep(2)
```
Comment (antban) [inline bubuku/features/restart_if_dead.py:40]: Please, do not use `sleep()` function, as it will block execution here (meaning that it will block all the functions of bubuku, including requests from cli interface)

File: bubuku/features/restart_if_dead.py [hunk 2]
```diff
@@ -18,18 +19,31 @@ def __init__(self, broker: BrokerManager, zk: BukuExhibitor):
...
 
         # Do not start if broker is running and registered
         def _cancel_if():
-            return self.broker.is_running_and_registered()
+            return self.is_running_and_registered()
 
         return RestartBrokerChange(self.zk, self.broker, _cancel_if, self.on_check_removed)
 
+    def is_running_and_registered(self, tries=1):
+        if not self.broker.process.is_running():
```
Comment (antban) [inline bubuku/features/restart_if_dead.py:35]: It is better to create 2 methods in broker class: `is_process_running()` and `is_broker_registered()`, so internals of broker (that are incorrectly named without underscore) wont' leak anywhere

File: bubuku/features/restart_if_dead.py (hunk 1 above)
Comment (adyach) [reply bubuku/features/restart_if_dead.py:40]: just to clarify a bit: the checks are executed with the certain interval by main loop / thread. sleeping here will block every check. the right approach is to count how many times broker was not registered, and if it is some number in a row, like 3 (3 times the check returned false in is_registered) then it should be really the case.

Comment (tor-vs-floki) [reply bubuku/features/restart_if_dead.py]: I reason that if a broker is not registered into Zookeeper, it makes sense to wait for an interval longer than the configured value of `zookeeper.connection.timeout.ms` until we can run any other change or check. 

The zookeeper client might succeed in restoring the session after a few attempts.

Ideally, we would configure Kafka to die on its own if the Zookeeper client cannot restore the session after a few attempts with an interval that we configured, and this check would only restart the process if it's not running.

PS: If the broker lost the connection with the Zookeeper and the session has died, does it make sense to run other checks or changes?

File: bubuku/features/restart_if_dead.py (hunk 2 above)
Comment (tor-vs-floki) [reply bubuku/features/restart_if_dead.py]: Agreed

File: bubuku/features/restart_if_dead.py [hunk 3]
```diff
@@ -14,22 +15,38 @@ def __init__(self, broker: BrokerManager, zk: BukuExhibitor):
...
 
         return RestartBrokerChange(self.zk, self.broker, _cancel_if, self.on_check_removed)
 
+    # Attempt to verify that broker is not registered in zookeeper for twice as long as the zookeeper session timeout.
+    # Allow zookeeper client to try to restore the session before killing tha kafka process as soon as zookeeper session is dead.
+    def is_running_and_registered(self, should_retry=False):
+        if not self.broker.is_running():
+            return False
+        if not self.broker.is_registered_in_zookeeper():
+            _LOG.warning('Broker is to not be regiestered in Zookeeper')
```
Comment (adyach) [inline bubuku/features/restart_if_dead.py:41]:
```suggestion
            _LOG.warning('Broker is not registered in Zookeeper')
```

File: bubuku/features/restart_if_dead.py [hunk 4]
```diff
@@ -14,22 +15,38 @@ def __init__(self, broker: BrokerManager, zk: BukuExhibitor):
...
+            time_to_return = self.last_zk_session_check + timedelta(milliseconds=self.broker.get_zookeeper_session_timeout() * 2)
+            if datetime.now() > time_to_return or not should_retry:
```
Comment (adyach) [inline bubuku/features/restart_if_dead.py:43]: 2 things:
- I would check for `should_retry` before doing any calculations
- I would extract datetime.now(). into variable, like current_time = datetime.now(), since it is used 3 times

**Comment by adyach** [comment]
👍

File: bubuku/features/restart_if_dead.py [hunk 5]
```diff
@@ -8,48 +8,47 @@
...
+        if not self.should_restart():
             return None
 
         _LOG.warning('Oops! Broker is dead, triggering restart')
         self.need_check = False
 
         # Do not start if broker is running and registered
         def _cancel_if():
-            return self.is_running_and_registered()
+            return not self.broker.is_running() or self.broker.is_registered_in_zookeeper()
```
Comment (antban) [inline bubuku/features/restart_if_dead.py:30]: Seems that the the line meaning has changed: 
was - cancel if broker is running and registered
now - cancel if broker is (not running) or registered

File: bubuku/features/restart_if_dead.py [hunk 6]
```diff
@@ -8,48 +8,47 @@
...
-            time_to_return = self.last_zk_session_check + timedelta(milliseconds=self.broker.get_zookeeper_session_timeout() * 2)
-            if current_time > time_to_return or not should_retry:
-                self.last_zk_session_check = current_time
-                return False
-        else:
-            self.last_zk_session_check = current_time
-        return True
+            if not self.last_zk_session_failed_check:
+                self.last_zk_session_failed_check = current_time
+            time_to_restart = self.last_zk_session_failed_check + timedelta(milliseconds=self.broker.get_zookeeper_session_timeout() * 2)
```
Comment (antban) [inline bubuku/features/restart_if_dead.py:44]: Maybe it is only a matter of taste, but I would prefer to name `time_to_restart` as `time_to_restart_at`

File: bubuku/features/restart_if_dead.py [hunk 7]
```diff
@@ -8,48 +8,47 @@
...
+            if current_time > time_to_restart:
+                return True
+        return False
 
     def on_check_removed(self):
```
Comment (antban) [inline bubuku/features/restart_if_dead.py:49]: I know that it is not part of this PR, but it will be better to call this method `on_change_removed`

Comment (tor-vs-floki) [reply bubuku/features/restart_if_dead.py:49]: Agreed, and what about `on_change_executed`

File: bubuku/features/restart_if_dead.py (hunk 5 above)
Comment (tor-vs-floki) [reply bubuku/features/restart_if_dead.py:30]: Oops

Comment (tor-vs-floki) [reply bubuku/features/restart_if_dead.py]: It should bail reboot when process is running and registered in zookeeper

**Comment by antban** [comment]
:+1:
